zkevm-circuits/src/tx_circuit.rs

```diff
@@ -0,0 +1,490 @@
+                            tx.to
+                                .unwrap_or_else(Address::zero)
+                                .to_scalar()
+                                .expect("tx.to too big"),
+                        ),
+                        (TxFieldTag::IsCreate, F::from(tx.to.is_none() as u64)),
+                        (
+                            TxFieldTag::Value,
+                            rlc(tx.value.to_le_bytes(), self.randomness),
+                        ),
+                        (
+                            TxFieldTag::CallDataLength,
+                            F::from(tx.call_data.0.len() as u64),
+                        ),
+                        (
+                            TxFieldTag::TxSignHash,
+                            *msg_hash_rlc_value.unwrap_or(&F::zero()),
+                        ),
+                    ] {
+                        let assigned_cell =
+                            config.assign_row(&mut region, offset, i + 1, *tag, 0, *value)?;
+                        offset += 1;
+
+                        // Ref. spec 0. Copy constraints using fixed offsets between the tx rows and
+                        // the SignVerifyChip
+                        match tag {
+                            TxFieldTag::CallerAddress => {
+                                region.constrain_equal(assigned_cell.cell(), address_cell)?
+                            }
+                            TxFieldTag::TxSignHash => {
+                                region.constrain_equal(assigned_cell.cell(), msg_hash_rlc_cell)?
+                            }
+                            _ => (),
+                        }
+                    }
+                }
+
+                // Assign call data
+                let mut calldata_count = 0;
+                for (i, tx) in self.txs.iter().enumerate() {
+                    for (index, byte) in tx.call_data.0.iter().enumerate() {
+                        assert!(calldata_count < MAX_CALLDATA);
+                        config.assign_row(
+                            &mut region,
+                            offset,
+                            i + 1, // tx_id
+                            TxFieldTag::CallData,
+                            index,
+                            F::from(*byte as u64),
+                        )?;
+                        offset += 1;
+                        calldata_count += 1;
+                    }
+                }
+                for _ in calldata_count..MAX_CALLDATA {
+                    config.assign_row(
+                        &mut region,
+                        offset,
+                        0, // tx_id
+                        TxFieldTag::CallData,
+                        0,
+                        F::zero(),
+                    )?;
+                    offset += 1;
+                }
+                Ok(())
+            },
+        )?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
```

Comment:
I think that we need an extra test here that is to check using using one already existing transaction (like a vector test).

Reply:
In https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/484/commits/4fc776d96d689b1d1e5d3959e8726295269d8891 I've added a test with a valid hardcoded tx, and another failure test with a hardcoded tx with invalid address.

Reply:
Nice! :)